From 975846fb3705148508f1a193dfc9f307995ccf65 Mon Sep 17 00:00:00 2001
From: prasannas <prasannas@ami.com>
Date: Fri, 13 Oct 2023 16:46:00 +0530
Subject: [PATCH 104/104] Fix for Display Hostname properly in NetworkProtocol

Description:
Code Fix to Display HostName Properly in Network Protocol.
Testcase:
1)Patching 64 Characters,HostName in FQDN Redfish
patch call
 https://x.x.x.x/redfish/v1/Managers/bmc/EthernetInterfaces/eth0
Body:
{
    "HostName":"AMIOT-f2b77ff5a01a1111111111111111111111111111111111111111111111"
}
Response:
  Response should be in 200Ok
2)Response in NetworkProtocol Uri
Get call
https://x.x.x.x/redfish/v1/Managers/bmc/NetworkProtocol
Response:
"HostName":"AMIOT-f2b77ff5a01a1111111111111111111111111111111111111111111111"

The hostname is displayed as per code changes.

Signed-off-by: prasannas <prasannas@ami.com>
---
 redfish-core/lib/ethernet.hpp         | 9 ++++-----
 redfish-core/lib/network_protocol.hpp | 2 +-
 2 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/redfish-core/lib/ethernet.hpp b/redfish-core/lib/ethernet.hpp
index aa4ea79..d3d03e4 100644
--- a/redfish-core/lib/ethernet.hpp
+++ b/redfish-core/lib/ethernet.hpp
@@ -902,12 +902,11 @@ inline bool isHostnameValid(const std::string& hostname)
         return false;
     }
     // Each label(hostname/subdomains) within a valid FQDN
-    // MUST handle host names of up to 63 characters (RFC 1123)
+    // MUST handle host names of up to 64 characters (RFC 1123)
     // labels cannot start or end with hyphens (RFC 952)
-    // labels can start with numbers (RFC 1123)
-    const static std::regex pattern(
-        "^[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]{0,61}[a-zA-Z0-9]$");
-
+    // hostname starts with an alphanumeric character
+    const std::regex pattern("^[a-zA-Z0-9]([a-zA-Z0-9\\-]{0,62}[a-zA-Z0-9])?$");
+ 
     return std::regex_match(hostname, pattern);
 }
 
diff --git a/redfish-core/lib/network_protocol.hpp b/redfish-core/lib/network_protocol.hpp
index 151f993..c4c655a 100644
--- a/redfish-core/lib/network_protocol.hpp
+++ b/redfish-core/lib/network_protocol.hpp
@@ -439,7 +439,7 @@ inline std::string getHostName()
 {
     std::string hostName;
 
-    std::array<char, HOST_NAME_MAX> hostNameCStr{};
+    std::array<char, HOST_NAME_MAX + 1> hostNameCStr{};
     if (gethostname(hostNameCStr.data(), hostNameCStr.size()) == 0)
     {
         hostName = hostNameCStr.data();
-- 
2.9.5

